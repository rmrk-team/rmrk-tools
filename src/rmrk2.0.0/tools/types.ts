import { Send } from "../classes/send";
import { Buy } from "../classes/buy";
import { List } from "../classes/list";
import { Emote } from "../classes/emote";
import { ChangeIssuer } from "../classes/changeissuer";
import { OP_TYPES } from "./constants";

export type IProperties = Record<string, IAttribute | IRoyaltyAttribute>;

export interface IAttribute {
  _mutation?: {
    allowed: boolean;
    with?: {
      opType: OP_TYPES;
      condition?: string;
    };
  };
  type:
    | "array"
    | "object"
    | "int"
    | "float"
    | "number"
    | "string"
    | "boolean"
    | "datetime"
    | "royalty";
  value: any;
}

export interface IRoyaltyAttribute extends IAttribute {
  type: "royalty";
  value: {
    receiver: string;
    royaltyPercentFloat: number;
  };
}

export interface Metadata {
  mediaUri?: string;
  thumbnailUri?: string;
  externalUri?: string;
  description?: string;
  name?: string;
  license?: string;
  licenseUri?: string;
  type?: string;
  locale?: string;
  properties?: IProperties;
  /** @deprecated deprecated in favour of `externalUri` field */
  external_url?: string;
  /** @deprecated deprecated in favour of `mediaUri` or `thumbnailUri` field */
  image?: string;
  /** @deprecated */
  image_data?: string;
  /** @deprecated deprecated in favour of `mediaUri` field */
  animation_url?: string;
}

export type Options = {
  ws: string;
  from: string;
  to: string;
  prefixes: string;
  blocks: string;
  json: string;
  folder: string;
  append: string;
  remark: string;
};

export type BlockCalls = {
  block: number;
  calls: BlockCall[];
};

export type BlockCall = {
  call: string;
  value: string;
  caller: string;
  extras?: BlockCall[];
};

export type BaseType = "svg" | "png" | "glb" | string;

export type Interaction = Send | Buy | List | Emote | ChangeIssuer;
