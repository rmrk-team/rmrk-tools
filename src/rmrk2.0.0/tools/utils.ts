import { ApiPromise, WsProvider } from "@polkadot/api";
import { hexToString, stringToHex } from "@polkadot/util";
import { URL } from "url";
import { Remark } from "./consolidator/remark";
import { OP_TYPES, VERSION } from "./constants";
import { SignedBlock } from "@polkadot/types/interfaces/runtime";
import { BlockCall, BlockCalls } from "./types";
import { Call as TCall } from "@polkadot/types/interfaces";
import { BlockHash } from "@polkadot/types/interfaces/chain";
import { encodeAddress } from "@polkadot/keyring";
import { deriveMultisigAddress } from "./deriveMultisigAddress";

export const getApi = async (wsEndpoint: string): Promise<ApiPromise> => {
  const wsProvider = new WsProvider(wsEndpoint);
  const api = ApiPromise.create({ provider: wsProvider });
  return api;
};

export const getLatestBlock = async (api: ApiPromise): Promise<number> => {
  const header = await api.rpc.chain.getHeader();
  return header.number.toNumber();
};

export const getLatestFinalizedBlock = async (
  api: ApiPromise
): Promise<number> => {
  const hash = await api.rpc.chain.getFinalizedHead();
  const header = await api.rpc.chain.getHeader(hash);
  if (header.number.toNumber() === 0) {
    console.error("Unable to retrieve finalized head - returned genesis block");
    process.exit(1);
  }
  return header.number.toNumber();
};

export const deeplog = function (obj: any): void {
  //@ts-ignore
  BigInt.prototype.toJSON = function () {
    return this.toString();
  };
  console.log(JSON.stringify(obj, null, 2));
};

export const stringIsAValidUrl = (s: string): boolean => {
  try {
    new URL(s);
    return true;
  } catch (err) {
    return false;
  }
};

export const prefixToArray = (prefix: string): string[] =>
  prefix.split(",").map((item) => {
    if (item.indexOf("0x") === 0) {
      return item;
    }
    return stringToHex(item);
  });

const getMeta = (call: Call, block: number): RemarkMeta | false => {
  const str = hexToString(call.value);
  const arr = str.split("::");
  if (arr.length < 3) {
    console.error(`Invalid RMRK in block ${block}: ${str}`);
    return false;
  }
  return {
    type: arr[1],
    version: parseFloat(arr[2]) ? arr[2] : "0.1",
  } as RemarkMeta;
};

type RemarkMeta = {
  type: string;
  version: string;
};

type Call = BlockCall;

export type Block = {
  block: number;
  calls: Call[];
};

export const getRemarksFromBlocks = (
  blocks: Block[],
  prefixes: string[],
  ss58Format = 2
): Remark[] => {
  const remarks: Remark[] = [];
  for (const row of blocks) {
    for (const call of row.calls) {
      if (call.call !== "system.remark") continue;
      const str = hexToString(call.value);
      const isValid = validateDecode(str);
      if (
        !isValid ||
        !prefixes.some((word) => str.startsWith(hexToString(word))) ||
        !str.includes(`::${VERSION}::`)
      ) {
        continue;
      }
      const meta = getMeta(call, row.block);
      if (!meta) continue;
      let remark;

      switch (meta.type) {
        case OP_TYPES.MINT:
        case OP_TYPES.CREATE:
        case OP_TYPES.RESADD:
        case OP_TYPES.THEMEADD:
        case OP_TYPES.SETPROPERTY:
        case OP_TYPES.SETPRIORITY:
        case OP_TYPES.BASE:
          remark = decodeURI(hexToString(call.value));
          break;
        default:
          remark = hexToString(call.value);
          break;
      }

      const r: Remark = {
        block: row.block,
        caller: encodeAddress(call.caller, ss58Format),
        interaction_type: meta.type,
        version: meta.version,
        remark: remark,
        extra_ex: call.extras,
      };
      remarks.push(r);
    }
  }
  return remarks;
};

export const isBatchInterrupted = async (
  api: ApiPromise,
  blockHash: BlockHash,
  extrinsicIndex: number
): Promise<boolean> => {
  const records = await api.query.system.events.at(blockHash);
  const events = records.filter(
    ({ phase, event }) =>
      phase.isApplyExtrinsic &&
      phase.asApplyExtrinsic.eq(extrinsicIndex) &&
      (event.method.toString() === "BatchInterrupted" ||
        event.method.toString() === "ExtrinsicFailed")
  );

  return Boolean(events.length);
};

export const validateDecode = (value: string) => {
  try {
    const decoded = decodeURI(value);
    return true;
  } catch (error: any) {
    console.log(error, value);
    return false;
  }
};

export const isSystemRemark = (call: TCall, prefixes: string[]): boolean =>
  call.section === "system" &&
  call.method === "remark" &&
  (prefixes.length < 1 ||
    prefixes.some((word) => call.args.toString().startsWith(word)));

export const isUtilityBatch = (call: TCall) =>
  call.section === "utility" &&
  (call.method === "batch" || call.method === "batchAll");

export const isMultiSig = (call: TCall) =>
  call.section === "multisig" && call.method === "asMulti";

export const getBlockCallsFromSignedBlock = async (
  signedBlock: SignedBlock,
  prefixes: string[],
  api: ApiPromise,
  ss58Format = 2
): Promise<BlockCall[] | []> => {
  let blockCalls: BlockCall[] = [];
  const extrinsics = signedBlock?.block?.extrinsics;
  if (!Array.isArray(extrinsics)) {
    return blockCalls;
  }

  let extrinsicIndex = 0;
  for (const extrinsic of extrinsics) {
    if (extrinsic.isEmpty || !extrinsic.isSigned) {
      extrinsicIndex++;
      continue;
    }

    if (isSystemRemark(extrinsic.method as TCall, prefixes)) {
      blockCalls.push({
        call: "system.remark",
        value: extrinsic.args.toString(),
        caller: encodeAddress(extrinsic.signer.toString(), ss58Format),
      });
    } else if (isMultiSig(extrinsic.method as TCall)) {
      /*
      First argument is multisig signers treshold, second is array of signer addresses,
      3rd some metadata and 4th can be a system.remark extrinsic call if one is sent
       */

      const [threshold, addresses, _, multisigRemarkHex] =
        extrinsic.method?.args || [];
      if (multisigRemarkHex) {
        try {
          const allMultisigAddresses = [
            ...(addresses.toJSON() as []),
            extrinsic.signer.toString(),
          ];
          const derivedMultisigAccount = deriveMultisigAddress({
            addresses: allMultisigAddresses,
            threshold: Number(threshold.toString()),
            ss58Prefix: ss58Format,
          });
          const multiSigRemarkCall = api.registry.createType(
            "Call",
            multisigRemarkHex.toU8a(true)
          );

          if (isSystemRemark(multiSigRemarkCall, prefixes)) {
            blockCalls.push({
              call: "system.remark",
              value: multiSigRemarkCall.args.toString(),
              caller: derivedMultisigAccount,
            });
          }
        } catch (error: any) {
          console.log(
            `Skipping multisig call ${signedBlock.block?.header?.number}-${extrinsicIndex} due to the fact that we cannot decode 3rd argument which is supposed to be system.remark`,
            error
          );
          extrinsicIndex++;
          continue;
        }
      }
    } else if (isUtilityBatch(extrinsic.method as TCall)) {
      // @ts-ignore
      const batchArgs: TCall[] = extrinsic.method.args[0];
      let remarkExists = 0;
      batchArgs.forEach((el) => {
        if (isSystemRemark(el, prefixes)) {
          remarkExists++;
        }
      });

      if (remarkExists) {
        const skip = await isBatchInterrupted(
          api,
          signedBlock.block.hash,
          extrinsicIndex
        );

        if (skip) {
          console.log(
            `Skipping batch ${signedBlock.block?.header?.number}-${extrinsicIndex} due to BatchInterrupted`
          );
          extrinsicIndex++;
          continue;
        }

        const batchRoot = [] as BlockCall[];
        const batchExtras: BlockCall[] = [];
        const batchBuyExtras: BlockCall[] = [];
        batchArgs.forEach((el, i) => {
          if (isSystemRemark(el, prefixes)) {
            batchRoot.push({
              call: `${el.section}.${el.method}`,
              value: el.args.toString(),
              caller: encodeAddress(extrinsic.signer.toString(), ss58Format),
            } as BlockCall);
          } else {
            const isBalanceTransfer =
              `${el.section}.${el.method}` === `balances.transfer`;

            const extraCall = {
              call: `${el.section}.${el.method}`,
              value: el.args.toString(),
              caller: encodeAddress(extrinsic.signer.toString(), ss58Format),
            } as BlockCall;

            if (isBalanceTransfer) {
              batchBuyExtras.push(extraCall);
            } else {
              batchExtras.push(extraCall);
            }
          }
        });

        if (batchExtras.length) {
          batchRoot.forEach((el, i) => {
            batchRoot[i].extras = batchExtras;
          });
        }

        // Only 1 NFT can be purchased in a batch
        if (batchBuyExtras.length) {
          batchRoot[0].extras = batchBuyExtras;
        }

        blockCalls = blockCalls.concat(batchRoot);
      }
    }
    extrinsicIndex++;
  }

  return blockCalls;
};

export const getRemarkData = (dataString: string) => {
  const data = decodeURIComponent(dataString);
  try {
    return JSON.parse(data);
  } catch (error) {
    return data;
  }
};

/**
 * filterBlocksByCollection
 * Return blocks that match specific collection.
 * @param blockCalls
 * @param collectionFilter - name of the collection to filter by
 */
export const filterBlocksByCollection = (
  blockCalls: BlockCalls[],
  prefixes: string[],
  collectionFilter?: string,
  ss58Format?: number
): BlockCalls[] =>
  blockCalls.filter((block) =>
    getRemarksFromBlocks([block], prefixes, ss58Format).some(
      (rmrk) =>
        (collectionFilter && rmrk.remark.includes(collectionFilter)) ||
        !collectionFilter
    )
  );
