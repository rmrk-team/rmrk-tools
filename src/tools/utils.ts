import { ApiPromise, WsProvider } from "@polkadot/api";
import { URL } from "url";

export const getApi = async (wsEndpoint: string): Promise<ApiPromise> => {
  const wsProvider = new WsProvider(wsEndpoint);
  const api = ApiPromise.create({ provider: wsProvider });
  return api;
};

export const getLatestBlock = async (api: ApiPromise): Promise<number> => {
  const header = await api.rpc.chain.getHeader();
  return header.number.toNumber();
};

export const getLatestFinalizedBlock = async (
  api: ApiPromise
): Promise<number> => {
  const hash = await api.rpc.chain.getFinalizedHead();
  const header = await api.rpc.chain.getHeader(hash);
  if (header.number.toNumber() === 0) {
    console.error("Unable to retrieve finalized head - returned genesis block");
    process.exit(1);
  }
  return header.number.toNumber();
};

export const deeplog = function (obj: any): void {
  console.log(JSON.stringify(obj, null, 2));
};

export const stringIsAValidUrl = (s: string): boolean => {
  try {
    new URL(s);
    return true;
  } catch (err) {
    return false;
  }
};
