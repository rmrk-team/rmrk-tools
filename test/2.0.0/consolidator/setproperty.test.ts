import { Consolidator, NFT } from "../../../src/rmrk2.0.0";
import {
  createCollectionMock,
  createCollectionMock2,
  getAliceKey,
  getBlockCallsMock,
  getBobKey,
  getRemarksFromBlocksMock,
  mintNftMock2,
} from "../mocks";
import { cryptoWaitReady } from "@polkadot/util-crypto";
import { OP_TYPES } from "../../../src/rmrk2.0.0/tools/constants";
import { stringToHex } from "@polkadot/util";

beforeAll(async () => {
  return await cryptoWaitReady();
});

const mintNftWithPropertiesOwner = (block?: number) =>
  new NFT({
    block: block || 0,
    collection: createCollectionMock().id,
    symbol: "KANR",
    sn: "777".padStart(8, "0"),
    transferable: 1,
    owner: getAliceKey().address,
    properties: {
      test: {
        _mutation: {
          allowed: true,
        },
        value: "foo",
        type: "string",
      },
    },
  });

const mintGemNft = (block?: number) =>
  new NFT({
    block: block || 0,
    collection: createCollectionMock2().id,
    symbol: "KANGEM",
    sn: "111".padStart(8, "0"),
    transferable: 1,
    owner: getAliceKey().address,
  });

const mintNftWithMutatorCondition = (block?: number) =>
  new NFT({
    block: block || 0,
    collection: createCollectionMock().id,
    symbol: "KANR",
    sn: "888".padStart(8, "0"),
    transferable: 1,
    owner: getAliceKey().address,
    properties: {
      test: {
        _mutation: {
          allowed: true,
          with: {
            opType: OP_TYPES.BURN,
            condition: "d43593c715a56da27d-KANARIAGEMS",
          },
        },
        value: "foo",
        type: "string",
      },
    },
  });

const mintNftWithProperties = (block?: number) =>
  new NFT({
    block: block || 0,
    collection: createCollectionMock().id,
    symbol: "KANR",
    sn: "777".padStart(8, "0"),
    transferable: 1,
    owner: getAliceKey().address,
    properties: {
      test: {
        value: "foo",
        type: "string",
      },
      royaltyInfo: {
        type: "royalty",
        value: {
          receiver: "xxx",
          royaltyPercentFloat: 0.2,
        },
        _mutation: {
          allowed: true,
        },
      },
    },
  });

describe("rmrk2.0.0 Consolidator: SETPROPERTY", () => {
  const getSetupRemarks = () => [
    ...getBlockCallsMock(createCollectionMock().create()),
    ...getBlockCallsMock(createCollectionMock2().create()),
  ];

  it("should throw if we are trying to mutate immutable property", async () => {
    expect(() => mintNftWithProperties(5).setproperty("test", 2)).toThrow();
  });

  it("should throw if we are trying to mutate property with wrong rootowner _mutation", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftWithPropertiesOwner().mint()),
      ...getBlockCallsMock(
        mintNftWithPropertiesOwner(4).setproperty("test", "bar"),
        getBobKey().address
      ),
    ]);
    const consolidator = new Consolidator();
    const consolidatedResult = await consolidator.consolidate(remarks);
    expect(consolidatedResult.invalid[0].message).toEqual(
      "[SETPROPERTY] Attempting to set property on a non-owned NFT. Expected FoQJpPyadYccjavVdTWxpxU7rUEaYhfLCPwXgkfD6Zat9QP but received HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F"
    );
  });

  it("should throw if we are trying to mutate property with no matching extra call passed", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftWithMutatorCondition().mint()),
      ...getBlockCallsMock(
        mintNftWithMutatorCondition(4).setproperty("test", "bar")
      ),
    ]);
    const consolidator = new Consolidator();
    const consolidatedResult = await consolidator.consolidate(remarks);
    expect(consolidatedResult.invalid[0].message).toEqual(
      "[SETPROPERTY] Attempting to mutate property without matching extra call of op type BURN"
    );
  });

  it("should throw if we are trying to mutate property with no matching extra call passed", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftWithMutatorCondition().mint()),
      ...getBlockCallsMock(
        mintNftWithMutatorCondition(4).setproperty("test", "bar"),
        undefined,
        [
          {
            call: "system.remark",
            value: stringToHex(mintNftWithMutatorCondition().mint()),
            caller: getAliceKey().address,
          },
        ]
      ),
    ]);
    const consolidator = new Consolidator();
    const consolidatedResult = await consolidator.consolidate(remarks);
    expect(consolidatedResult.invalid[0].message).toEqual(
      "The op code needs to be BURN, but it is MINT"
    );
  });

  it("should throw if we are trying to mutate property with no matching extra call condition", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftWithMutatorCondition().mint()),
      ...getBlockCallsMock(mintNftWithPropertiesOwner().mint()),
      ...getBlockCallsMock(
        mintNftWithMutatorCondition(4).setproperty("test", "bar"),
        undefined,
        [
          {
            call: "system.remark",
            value: stringToHex(mintNftWithPropertiesOwner(5).burn()),
            caller: getAliceKey().address,
          },
        ]
      ),
    ]);
    const consolidator = new Consolidator();
    const consolidatedResult = await consolidator.consolidate(remarks);
    expect(consolidatedResult).toMatchSnapshot();
  });

  it("should allow if we are trying to mutate property with matching mutation condition", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftWithMutatorCondition().mint()),
      ...getBlockCallsMock(mintGemNft().mint()),
      ...getBlockCallsMock(
        mintNftWithMutatorCondition(4).setproperty("test", "bar"),
        undefined,
        [
          {
            call: "system.remark",
            value: stringToHex(mintGemNft(5).burn()),
            caller: getAliceKey().address,
          },
        ]
      ),
    ]);
    const consolidator = new Consolidator();
    const consolidatedResult = await consolidator.consolidate(remarks);
    expect(consolidatedResult).toMatchSnapshot();
  });

  it("should allow to mutate property if owner _mutator match", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftWithPropertiesOwner().mint()),
      ...getBlockCallsMock(
        mintNftWithPropertiesOwner(4).send(getBobKey().address)
      ), // Send to Bob first
      ...getBlockCallsMock(
        mintNftWithPropertiesOwner(4).setproperty("test", "bar"),
        getBobKey().address
      ),
    ]);
    const consolidator = new Consolidator();
    const consolidatedResult = await consolidator.consolidate(remarks);
    expect(consolidatedResult).toMatchSnapshot();
  });

  it("should be able to freeze property ", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftWithPropertiesOwner().mint()),
      ...getBlockCallsMock(
        mintNftWithPropertiesOwner(4).setproperty("test", "bar", "freeze")
      ),
      ...getBlockCallsMock(
        mintNftWithPropertiesOwner(4).setproperty("test", "baz")
      ),
    ]);
    const consolidator = new Consolidator();
    const consolidatedResult = await consolidator.consolidate(remarks);
    expect(consolidatedResult.invalid[0].message).toEqual(
      "[SETPROPERTY] Attempting to set property on immutable property test"
    );
  });

  it("should allow to mutate royalty if owner is an issuer", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftWithProperties().mint()),
      ...getBlockCallsMock(
        mintNftWithProperties(4).setproperty("royaltyInfo", {
          receiver: "xxx",
          royaltyPercentFloat: 0.3,
        })
      ),
    ]);
    const consolidator = new Consolidator();
    const consolidatedResult = await consolidator.consolidate(remarks);
    expect(
      consolidatedResult.nfts[mintNftWithProperties(4).getId()].properties
        ?.royaltyInfo?.value?.royaltyPercentFloat
    ).toEqual(0.3);
  });

  it("should not allow to mutate royalty ", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftWithProperties().mint(getBobKey().address)),
      ...getBlockCallsMock(
        mintNftWithProperties(4).setproperty("royaltyInfo", {
          receiver: "xxx",
          royaltyPercentFloat: 0.3,
        }),
        getBobKey().address
      ),
    ]);
    const consolidator = new Consolidator();
    const consolidatedResult = await consolidator.consolidate(remarks);
    expect(consolidatedResult.invalid[0].message).toEqual(
      "[SETPROPERTY] Only issuer can mutate an attribute of type 'royalty'."
    );
  });

  it("should not allow to mutate royalty to an invalid value", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftWithProperties().mint()),
      ...getBlockCallsMock(
        mintNftWithProperties(4).setproperty("royaltyInfo", {
          receiver: "xxx",
          royaltyPercentFloat: 101,
        })
      ),
    ]);
    const consolidator = new Consolidator();
    const consolidatedResult = await consolidator.consolidate(remarks);
    expect(consolidatedResult).toMatchSnapshot();
  });
});
