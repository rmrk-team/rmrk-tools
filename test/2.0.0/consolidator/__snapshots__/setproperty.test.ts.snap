// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rmrk2.0.0 Consolidator: SETPROPERTY should allow if we are trying to mutate property with matching mutation condition 1`] = `
Object {
  "bases": Object {},
  "collections": Object {
    "d43593c715a56da27d-KANARIABIRDS": Collection {
      "block": 2,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIABIRDS",
      "issuer": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIABIRDS",
    },
    "d43593c715a56da27d-KANARIAGEMS": Collection {
      "block": 3,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIAGEMS",
      "issuer": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIAGEMS",
    },
  },
  "invalid": Array [],
  "nfts": Object {
    "4-d43593c715a56da27d-KANARIABIRDS-KANR-00000888": Object {
      "block": 4,
      "burned": "",
      "changes": Array [],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "4-d43593c715a56da27d-KANARIABIRDS-KANR-00000888",
      "metadata": undefined,
      "owner": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "pending": false,
      "priority": Array [],
      "properties": Object {
        "test": Object {
          "_mutation": Object {
            "allowed": true,
            "with": Object {
              "condition": "d43593c715a56da27d-KANARIAGEMS",
              "opType": "BURN",
            },
          },
          "type": "string",
          "value": "bar",
        },
      },
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "sn": "00000888",
      "symbol": "KANR",
      "transferable": 1,
    },
    "5-d43593c715a56da27d-KANARIAGEMS-KANGEM-00000111": Object {
      "block": 5,
      "burned": "",
      "changes": Array [],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIAGEMS",
      "forsale": 0n,
      "id": "5-d43593c715a56da27d-KANARIAGEMS-KANGEM-00000111",
      "metadata": undefined,
      "owner": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "pending": false,
      "priority": Array [],
      "properties": Object {},
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "sn": "00000111",
      "symbol": "KANGEM",
      "transferable": 1,
    },
  },
}
`;

exports[`rmrk2.0.0 Consolidator: SETPROPERTY should allow to mutate property if owner _mutator match 1`] = `
Object {
  "bases": Object {},
  "collections": Object {
    "d43593c715a56da27d-KANARIABIRDS": Collection {
      "block": 2,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIABIRDS",
      "issuer": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIABIRDS",
    },
    "d43593c715a56da27d-KANARIAGEMS": Collection {
      "block": 3,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIAGEMS",
      "issuer": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIAGEMS",
    },
  },
  "invalid": Array [],
  "nfts": Object {
    "4-d43593c715a56da27d-KANARIABIRDS-KANR-00000777": Object {
      "block": 4,
      "burned": "",
      "changes": Array [
        Object {
          "block": 5,
          "caller": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
          "field": "owner",
          "new": "FoQJpPyadYccjavVdTWxpxU7rUEaYhfLCPwXgkfD6Zat9QP",
          "old": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
          "opType": "SEND",
        },
      ],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "4-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "metadata": undefined,
      "owner": "FoQJpPyadYccjavVdTWxpxU7rUEaYhfLCPwXgkfD6Zat9QP",
      "pending": false,
      "priority": Array [],
      "properties": Object {
        "test": Object {
          "_mutation": Object {
            "allowed": true,
          },
          "type": "string",
          "value": "bar",
        },
      },
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "FoQJpPyadYccjavVdTWxpxU7rUEaYhfLCPwXgkfD6Zat9QP",
      "sn": "00000777",
      "symbol": "KANR",
      "transferable": 1,
    },
  },
}
`;

exports[`rmrk2.0.0 Consolidator: SETPROPERTY should not allow to mutate royalty to an invalid value 1`] = `
Object {
  "bases": Object {},
  "collections": Object {
    "d43593c715a56da27d-KANARIABIRDS": Collection {
      "block": 2,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIABIRDS",
      "issuer": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIABIRDS",
    },
    "d43593c715a56da27d-KANARIAGEMS": Collection {
      "block": 3,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIAGEMS",
      "issuer": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIAGEMS",
    },
  },
  "invalid": Array [],
  "nfts": Object {
    "4-d43593c715a56da27d-KANARIABIRDS-KANR-00000777": Object {
      "block": 4,
      "burned": "",
      "changes": Array [],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "4-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "metadata": undefined,
      "owner": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "pending": false,
      "priority": Array [],
      "properties": Object {
        "royaltyInfo": Object {
          "_mutation": Object {
            "allowed": true,
          },
          "type": "royalty",
          "value": Object {
            "receiver": "xxx",
            "royaltyPercentFloat": 101,
          },
        },
        "test": Object {
          "type": "string",
          "value": "foo",
        },
      },
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "sn": "00000777",
      "symbol": "KANR",
      "transferable": 1,
    },
  },
}
`;

exports[`rmrk2.0.0 Consolidator: SETPROPERTY should throw if we are trying to mutate property with no matching extra call condition 1`] = `
Object {
  "bases": Object {},
  "collections": Object {
    "d43593c715a56da27d-KANARIABIRDS": Collection {
      "block": 2,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIABIRDS",
      "issuer": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIABIRDS",
    },
    "d43593c715a56da27d-KANARIAGEMS": Collection {
      "block": 3,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIAGEMS",
      "issuer": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIAGEMS",
    },
  },
  "invalid": Array [
    Object {
      "block": 6,
      "caller": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "message": "[SETPROPERTY] Attempting to mutate an property but it's mutation condition doesn't match",
      "object_id": "4-d43593c715a56da27d-KANARIABIRDS-KANR-00000888",
      "op_type": "SETPROPERTY",
    },
  ],
  "nfts": Object {
    "4-d43593c715a56da27d-KANARIABIRDS-KANR-00000888": Object {
      "block": 4,
      "burned": "",
      "changes": Array [],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "4-d43593c715a56da27d-KANARIABIRDS-KANR-00000888",
      "metadata": undefined,
      "owner": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "pending": false,
      "priority": Array [],
      "properties": Object {
        "test": Object {
          "_mutation": Object {
            "allowed": true,
            "with": Object {
              "condition": "d43593c715a56da27d-KANARIAGEMS",
              "opType": "BURN",
            },
          },
          "type": "string",
          "value": "foo",
        },
      },
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "sn": "00000888",
      "symbol": "KANR",
      "transferable": 1,
    },
    "5-d43593c715a56da27d-KANARIABIRDS-KANR-00000777": Object {
      "block": 5,
      "burned": "",
      "changes": Array [],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "5-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "metadata": undefined,
      "owner": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "pending": false,
      "priority": Array [],
      "properties": Object {
        "test": Object {
          "_mutation": Object {
            "allowed": true,
          },
          "type": "string",
          "value": "foo",
        },
      },
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "HNZata7iMYWmk5RvZRTiAsSDhV8366zq2YGb3tLH5Upf74F",
      "sn": "00000777",
      "symbol": "KANR",
      "transferable": 1,
    },
  },
}
`;
